R/broom_wrapper.R

```diff
@@ -205,31 +230,61 @@ prediction <- function(df, source_data, test = TRUE){
 
   # drop unnecessary columns
   df <- dplyr::select(df, .test_index, model)
+  cll <- match.call()
+  aug_args <- expand_args(cll, exclude = c("df", "source_data", "test"))
 
   ret <- if(test){
     # augment by test data
-    dplyr::bind_cols(df, source) %>%
+    # use formula to support expanded aug_args (especially for type.predict for logistic regression)
+    aug_fml <- if(aug_args == ""){
+      as.formula(paste0("~list(broom::augment(model, newdata = data))"))
+    } else {
+      as.formula(paste0("~list(broom::augment(model, newdata = data, ", aug_args, "))"))
+    }
+    data_to_augment <- dplyr::bind_cols(df, source) %>%
       dplyr::ungroup() %>%
       dplyr::mutate(data = purrr::map2(data, .test_index, function(df, index){
         # keep data only in test_index
         safe_slice(df, index)
       })) %>%
-      dplyr::mutate(data = purrr::map2(data, model, function(df, model){
-        # remove rows that have categories that aren't in training data
-        # otherwise, broom::augment causes an error
-        filtered_data <- df
-        for (cname in colnames(model$model)) {
-          filtered_data <- filtered_data[filtered_data[[cname]] %in% model$model[[cname]], ]
-        }
-        filtered_data
-      })) %>%
-      dplyr::select(-.test_index) %>%
-      dplyr::rowwise() %>%
-      dplyr::mutate(data = list(broom::augment(model, newdata = data))) %>%
+      dplyr::select(-.test_index)
+
+
+    augmented <- tryCatch({
+      data_to_augment %>%
+        dplyr::rowwise() %>%
+        dplyr::mutate_(.dots = list( data = aug_fml))
+    }, error = function(e){
+      if (grepl("arguments imply differing number of rows: ", e$message)) {
+        data_to_augment %>%
+          dplyr::mutate(data = purrr::map2(data, model, function(df, model){
+            # remove rows that have categories that aren't in training data
+            # otherwise, broom::augment causes an error
+            filtered_data <- df
+            for (cname in colnames(model$model)) {
+              filtered_data <- filtered_data[filtered_data[[cname]] %in% model$model[[cname]], ]
+            }
+            filtered_data
+          })) %>%
+          dplyr::rowwise() %>%
+          dplyr::mutate_(.dots = list( data = aug_fml))
+      } else {
+        stop(e$message)
+      }
+    })
+
+    ret <- augmented %>%
       dplyr::select(-model) %>%
       tidyr::unnest(data)
+
   } else {
     # augment by trainig data
+    # use formula to support expanded aug_args (especially for type.predict for logistic regression)
```

Comment:
This part looks same as the case test is TRUE. Can it be outside of if statement to avoid duplicate?